### ISSUE TITLE
When removing resources: cycle error during apply, but not during plan

### ISSUE THREAD

**Issue by stazz** [hashicorp/terraform#22561]
### Terraform Version

```
Terraform v0.12.6
```

### Terraform Configuration Files

Configuration is located here: https://gist.github.com/stazz/0dc6c79fbbadfbebae4508bfbc6595f7/raw/f7423172b7aac32bca27ef0e4cf96e0077564fab/config.zip .

### Debug Output

Debug output is located here: https://gist.github.com/stazz/0dc6c79fbbadfbebae4508bfbc6595f7/raw/3688c4c0607dc2f69e8bf5718431e6b1bd21687b/tf.log
### Expected Behavior

Apply should have succeeded.
### Actual Behavior

The apply command printed:
```
module.all-rgs.azurerm_resource_group.rgs["tfrepro-nw"]: Refreshing state... [id=/subscriptions/3f3f5192-6647-49a5-93cb-6ba1862a196f/resourceGroups/tfrepro-nw]
module.vnets.azurerm_virtual_network.vnet["one"]: Refreshing state... [id=/subscriptions/3f3f5192-6647-49a5-93cb-6ba1862a196f/resourceGroups/tfrepro-nw/providers/Microsoft.Network/v
irtualNetworks/tfrepro-vnet]

Error: Cycle: module.vnets.azurerm_virtual_network.vnet (prepare state), module.vnets.azurerm_virtual_network.vnet["one"] (destroy), module.all-rgs.azurerm_resource_group.rgs["tfrep
ro-nw"] (destroy), module.all-rgs.output.rgs, module.vnets.var.vnets
```
### Steps to Reproduce

1. Extract the zip file to some folder, e.g. `/tfrepro`, and open command line to `/tfrepro/code` folder.
2. Run `terraform init`, followed by `terraform apply -state=/tfrepro/terraform.tfstate -auto-approve -var-file=/tfrepro/vars/main.tfvars`. Observe the successful deployment of resource group and VNET.
3. Run `terraform plan -state=/tfrepro/terraform.tfstate` (notice leaving out `-var-file`). Observe command to **succeed**, and print a plan to remove the VNET and resource group, as the variable file is no longer given.
4. Run `terraform apply -state=/tfrepro/terraform.tfstate -auto-approve`. Observe the **error message** about the cycle.

### Additional Context

I ran the Terraform within the Docker image, as that's how it is run in the current project I am in. I used `0.12.6` tag from DockerHub: https://hub.docker.com/r/hashicorp/terraform/tags .

The `code` folder of the configuration `config.zip` file is a very simplified version of the Terraform code I currently use, which is not publicly available. The `rg` submodule is meant to create resource groups on-demand as specified by configuration.

The log output has been redacted of most of the output from `plugin.terraform-provider-azurerm`, as it contains sensitive values and is most likely not relevant to this problem, It looks to be more related to the internal graph analysis done by Terraform core.

### References

I tried to search, but none were very close to this issue.

**Comment by eddytrex** [comment]
Hi, maybe is related to this, https://github.com/hashicorp/terraform/issues/22502?

**Comment by stazz** [comment]
Hey @eddytrex , it might be so, thank you for pointing me to the issue you submitted! It is indeed very much so looking like this issue is duplicate of #22502 , however I am not very familiar with Terraform internals so I am not sure about that.

I need to test and verify if `create_before_destroy` would resolve this particular scenario, I'll do that when I get to office tomorrow.

**Comment by stazz** [comment]
Well, I tried various combinations of `create_before_destroy`: to the resource group, to the VNET, and then to both resource group and VNET. None of them helped with working around this specific issue tho.

Possibly a bit off-topic:

One interesting thing is that I also tried changing the key of the `vnet_info` variable (e.g. from `"one"` to `"two"`), and then running `apply`. It typically first produced the following error:
```
Error: Provider produced inconsistent result after apply

When applying changes to module.vnets.azurerm_virtual_network.vnet["two"],
provider "azurerm" produced an unexpected new value for was present, but now
absent.

This is a bug in the provider, which should be reported in the provider's own
issue tracker.
```

And when just running same command again, it re-created the VNET. Is this truly something I should report to provider's own issue tracker?

However, *occasionally* changing the key of `vnet_info` mapping and then running `apply` completed successfully right away! However, it resulted in claiming that both resource group **and** VNET exist, but when I checked from the Azure Portal, only the resource group existed. So it ended up in inconsistent state with the actual state in Azure. In this situation, the step 4 of reproduction steps of this issue actually completed successfully, but obviously the real problem in that case is that the starting state was not correct anyways. Is this also something I should report to provider's own issue tracker?

**Comment by stazz** [comment]
Just tested with `0.12.7`, still happens in that version as well.

**Comment by stazz** [comment]
Still happens with `0.12.10` as well, let's hope #22976 will help this.

**Comment by dmrzzz** [comment]
I'm seeing a cycle error which may or may not be the same underlying issue, but it perfectly matches this issue's _title_ so I'll comment here instead of opening a separate bug.

main.tf:
```hcl
terraform {
  required_version = ">= 0.12.12"
}

variable "region" {
  default = "us-east-2"
}

locals {
  gateway_vpc_endpoint_service_names = [
    "com.amazonaws.${var.region}.dynamodb",
    "com.amazonaws.${var.region}.s3",
  ]

  gateway_vpc_endpoint_ids = { for k in local.gateway_vpc_endpoint_service_names: k => random_pet.aws_vpc_endpoint[k].id }
}

resource "random_pet" "aws_vpc_endpoint" {
  for_each = toset(local.gateway_vpc_endpoint_service_names)
  prefix = each.value
}

module "subnet" {
  source = "./subnet"
  gateway_vpc_endpoint_ids = local.gateway_vpc_endpoint_ids
}
```

subnet/module.tf:
```hcl
variable "gateway_vpc_endpoint_ids" {
  type = map(string)
}

resource "random_pet" "aws_vpc_endpoint_route_table_association" {
  for_each = var.gateway_vpc_endpoint_ids
  prefix = each.value
}
```

Steps:

1. terraform-0.12.12 apply
    Creates two "endpoint" pets and two "association" pets as expected.

2. Edit main.tf and comment out the list item containing "dynamodb"

**Comment by dmrzzz** [comment]
TF 0.12.13 solves my use case.

**Comment by stazz** [comment]
I confirm that this is indeed fixed by `0.12.13`. I guess the #22976 PR fixed it. In any case, this is great news! :) Closing this now, as things are now working as expected in `0.12.13`.
